### PR TITLE
Assay improvements - issues #369, #370

### PR DESCRIPTION
**Pull request by Zigur**
Work in progress, for now, issue #369 is completed, #370 still todo.

### CODE REVIEW

**Comment by Zigur** [comment]
This is now ready to be reviewed and merged. @proccaserra have a look at the outputted ISA tables using https://github.com/Zigur/isa-create-notebook/blob/dev/notebooks/Investigation-from-datascriptor-config.ipynb for instance. Check that there are no duplicates in the node names (I'll add some tests on that too, before merging).

File: isatools/create/model.py [hunk 1]
```diff
@@ -2240,7 +2267,93 @@ def _generate_assay(assay_graph, assay_samples, cell_name=''):
...
+                        else technology_type
+                    curr_assay_opt = next(
+                        opt for opt in assays_opts if opt['measurement type'] == m_type_term and
+                        opt['technology type'] == t_type_term
+                    )
+                    log.debug('Assay conf. found: {}; {}; {};'.format(
+                        measurement_type, technology_type, curr_assay_opt)
+                    )
+                    isa_class = globals()[curr_assay_opt['raw data file'].replace(' ', '')]
+                    assert isa_class in {RawDataFile, RawSpectralDataFile}
```
Comment (proccaserra) [inline isatools/create/model.py:2337]: at least add the following key "'FreeInductionDecayDataFile"  (and possibly "ImageDataFile") 
TODO: review assay definitions and test all current layouts)

**Review by proccaserra — CHANGES_REQUESTED** [review]
Line1056 in create/model.py causes an exception "TypeError: expected string or bytes-like object" when the assay definition used is annotated with ontology terms (to reproduce switch flag for MS (annotated) to true, when testing with the notebook).

then at line 2334 in the same module, some required keys are missing "'FreeInductionDecayDataFile" (and possibly "ImageDataFile") - 
TODO: review assay definitions and test all current layouts)

File: isatools/create/model.py (hunk 1 above)
Comment (Zigur) [reply isatools/create/model.py]: This should be fixed now

**Comment by Zigur** [comment]
This is mostly completed, I'd say. I have included support for all data file types, both raw and derived, for completeness:

https://github.com/ISA-tools/isa-api/blob/746fde3f67aafc0f0a9cc0e3014ea4854c035baf/isatools/create/model.py#L2349-L2355

I have added two prefixes to the node names to clarify the notations:
- `AT` is prepended to the assay type id
- `S` is prepended to the start node (or sample node) index
- `R` is prepended to the replicate node index

Some examples:
- extract name: `AT1-S9-Extract-R1`
- labelled extract name: `AT1-S9-LE-R1`
- process name:  `AT1-S9-mass-spectrometry-Acquisition-R1`
- raw data file name:  `AT1-S9-raw-spectral-data-file-R1`

@proccaserra have a look at [this notebook](https://github.com/Zigur/isa-create-notebook/blob/dev/notebooks/Investigation-from-datascriptor-config-observational-variables-and-ontology-annotation.ipynb) for reviewing a scenario with the ontology annotated MS assay and the NMR assay

**Comment by Zigur** [comment]
This seems to work also for assays with chained protocols now (see [this](https://github.com/Zigur/isa-create-notebook/blob/dev/notebooks/Investigation-from-datascriptor-config.ipynb) notebook, at the moment the only functional one, others to be updated soon).

**Comment by proccaserra** [comment]
looks good to me now.

**Pull request merged**